cardano-node/src/Cardano/Node/Run.hs

```diff
@@ -325,31 +357,60 @@ handleSimpleNode p trace nodeTracers npm onKernel = do
 
          when mockValidateDB $ traceWith tracer "Performing DB validation"
 
+  stopOnSlotSynced :: NodeProtocolMode -> MaxSlotNo
+  stopOnSlotSynced (RealProtocolMode cli) = shutdownOnSlotSynced cli
+  stopOnSlotSynced _ = NoMaxSlotNo
+
+  -- | Produce a potential pair of FDs:
+  --  - read end, where EOF is a signal of termination
+  --  - write end, which is means of internal termination handling
+  decideShutdownFds :: NodeProtocolMode -> IO ShutdownFDs
+  decideShutdownFds (RealProtocolMode NodeCLI{shutdownIPC = Just fd}) = pure $ ExternalShutdown fd
+  decideShutdownFds (RealProtocolMode NodeCLI{shutdownOnSlotSynced = MaxSlotNo{}}) = do
+    (r, w) <- Posix.createPipe
+    pure $ InternalShutdown r w
+  decideShutdownFds _ = pure NoShutdownFDs
+
+data ShutdownFDs
+  = NoShutdownFDs
+  | ExternalShutdown Fd
+  -- ^ We shall watch this externally-supplied FD for termination EOF.
+  | InternalShutdown Fd Fd
+  -- ^ Pair of FDs for intra-process graceful shutdown termination messaging.
+  --   First FD is the read side as above,
+  --   second FD is the shutdown doorbell to be closed.
```

Comment:
Perhaps the choice of the name is unfortunate, as it reflects content, not intent.

Couldn't think of a more fitting name, though.  Maybe `ShutdownHandle`, but that's probably not much better..